network/gossip/libp2p/test/epochtransition_test.go

```diff
@@ -0,0 +1,218 @@
+package test
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	golog "github.com/ipfs/go-log"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/stretchr/testify/assert"
+	mock2 "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/network/gossip/libp2p"
+	protocol "github.com/onflow/flow-go/state/protocol/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type EpochTransitionTestSuite struct {
+	suite.Suite
+	ConduitWrapper
+	nets         []*libp2p.Network
+	mws          []*libp2p.Middleware
+	engines      []*MeshEngine
+	state        *protocol.ReadOnlyState
+	snapshot     *protocol.Snapshot
+	ids          flow.IdentityList
+	currentEpoch int //counter to track the current epoch
+	logger       zerolog.Logger
+}
+
+func TestEpochTransitionTestSuite(t *testing.T) {
+	suite.Run(t, new(EpochTransitionTestSuite))
+}
+
+func (ts *EpochTransitionTestSuite) SetupTest() {
+	nodeCount := 10
+	golog.SetAllLoggers(golog.LevelDebug)
+	ts.logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr}).With().Caller().Logger()
+
+	// create ids
+	ids, mws, err := generateIDsAndMiddlewares(nodeCount, ts.logger)
+	require.NoError(ts.T(), err)
+	ts.ids = ids
+	ts.mws = mws
+
+	// setup state related mocks
+	ts.state = new(protocol.ReadOnlyState)
+	ts.snapshot = new(protocol.Snapshot)
+	ts.state.On("Final").Return(ts.snapshot, nil)
+
+	ts.currentEpoch = 0
+	ts.updateSnapshot(0, ts.ids)
+
+	// all nodes use the same state mock
+	states := make([]*protocol.ReadOnlyState, nodeCount)
+	for i := 0; i < nodeCount; i++ {
+		states[i] = ts.state
+	}
+
+	// create networks using the mocked state and default topology
+	nets, err := generateNetworks(ts.logger, ids, mws, 100, nil, states, false)
+	require.NoError(ts.T(), err)
+	ts.nets = nets
+
+	// generate the engines
+	ts.engines = generateEngines(ts.T(), nets)
+
+	// ensure that nodes can communicate with each other
+	sendMessagesAndVerify(ts.T(), ts.ids, ts.engines, ts.Publish)
+}
+
+// TearDownTest closes the networks within a specified timeout
+func (ts *EpochTransitionTestSuite) TearDownTest() {
+	for _, net := range ts.nets {
+		select {
+		// closes the network
+		case <-net.Done():
+			continue
+		case <-time.After(3 * time.Second):
+			ts.Suite.Fail("could not stop the network")
+		}
+	}
+}
+
+// TestNewNodeAdded tests that an additional node in a new epoch get connected to other nodes and can exchange messages
+func (ts *EpochTransitionTestSuite) TestNewNodeAdded() {
+	ts.T().Skip("skipping till discovery is updated to discover newly added nodes")
```

Comment:
these test fail since currently there is no way to trigger the discovery on demand. Once https://github.com/dapperlabs/flow-go/issues/4893 is implemented, these should pass.